### PR TITLE
tailcfg: add field to allow LocalPortForwarding in SSHAction

### PR DESCRIPTION
**Pull request by maisem**
Updates #3802, #4129

Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

File: ssh/tailssh/tailssh.go [hunk 1]
```diff
@@ -96,16 +96,17 @@ type server struct {
...
 
-// portForward reports whether the ctx should be allowed to port forward
+// localPortForward reports whether the ctx should be allowed to port forward
 // to the specified host and port.
 // TODO(bradfitz/maisem): should we have more checks on host/port?
-func (srv *server) portForward(ctx ssh.Context, destinationHost string, destinationPort uint32) bool {
-	return srv.isActiveSession(ctx)
+func (srv *server) localPortForward(ctx ssh.Context, destinationHost string, destinationPort uint32) bool {
+	a := srv.sessionAction(ctx)
+	return a.LocalPortForwarding
```
Comment (bradfitz) [inline ssh/tailssh/tailssh.go:109]: I was going to say your API change on line 313 below was prone to misuse and this here looks what I was afraid of: nil ptr derefs after blind use trusting it'll return the right thing.

File: tailcfg/tailcfg.go [hunk 2]
```diff
@@ -1624,6 +1624,10 @@ type SSHAction struct {
 	// response, it should be re-fetched as long as the SSH
 	// session is open.
 	HoldAndDelegate string `json:"holdAndDelegate,omitempty"`
+
+	// LocalPortForwarding, if true, allows accepted connections
+	// to optionally use local port forwarding.
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:1629]: to use local port forwarding if requested.

?

File: tailcfg/tailcfg.go [hunk 3]
```diff
@@ -1624,6 +1624,10 @@ type SSHAction struct {
...
+	LocalPortForwarding bool `json:"localPortForwarding,omitempty"`
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:1630]: How about "AllowLocalPortForwarding" ?

File: ssh/tailssh/tailssh.go (hunk 1 above)
Comment (maisem) [reply ssh/tailssh/tailssh.go]: Sigh, yes. Fixed

File: tailcfg/tailcfg.go (hunk 3 above)
Comment (maisem) [reply tailcfg/tailcfg.go]: sure

File: tailcfg/tailcfg.go (hunk 2 above)
Comment (maisem) [reply tailcfg/tailcfg.go]: sure

File: ssh/tailssh/tailssh.go [hunk 4]
```diff
@@ -96,16 +96,17 @@ type server struct {
...
 	// mu protects activeSessions.
 	mu             sync.Mutex
-	activeSessions map[string]bool
+	activeSessions map[string]*tailcfg.SSHAction
 }
 
 var debugPolicyFile = envknob.String("TS_DEBUG_SSH_POLICY_FILE")
 
-// portForward reports whether the ctx should be allowed to port forward
+// localPortForward reports whether the ctx should be allowed to port forward
```
Comment (bradfitz) [inline ssh/tailssh/tailssh.go:104]: doesn't sound like a boolean func.

how about `mayForwardLocalPortTo`

File: ssh/tailssh/tailssh.go [hunk 5]
```diff
@@ -96,16 +96,17 @@ type server struct {
...
 // to the specified host and port.
 // TODO(bradfitz/maisem): should we have more checks on host/port?
-func (srv *server) portForward(ctx ssh.Context, destinationHost string, destinationPort uint32) bool {
-	return srv.isActiveSession(ctx)
+func (srv *server) localPortForward(ctx ssh.Context, destinationHost string, destinationPort uint32) bool {
```
Comment (bradfitz) [inline ssh/tailssh/tailssh.go:107]: I don't think this method should be on `*server`.

I think we need a new type for an established session, that `launchProcess` or something earlier should return.

File: ssh/tailssh/tailssh.go [hunk 6]
```diff
@@ -256,7 +257,7 @@ ProcessAction:
 		})
 		defer t.Stop()
 	}
-	srv.handleAcceptedSSH(ctx, s, ci, lu)
+	srv.handleAcceptedSSH(ctx, s, action, ci, lu)
```
Comment (bradfitz) [inline ssh/tailssh/tailssh.go:260]: yeah, this suggests we should have a new type before this, and then `handleAcceptedSSH` can be renamed to `run` or something on that new type.

File: tailcfg/tailcfg.go [hunk 7]
```diff
@@ -1624,6 +1624,10 @@ type SSHAction struct {
 	// response, it should be re-fetched as long as the SSH
 	// session is open.
 	HoldAndDelegate string `json:"holdAndDelegate,omitempty"`
+
+	// AllowLocalPortForwarding, if true, allows accepted connections
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:1628]: this part in this file LGTM though

**Review by bradfitz — CHANGES_REQUESTED** [review]

**Review by bradfitz — APPROVED** [review]

**Pull request merged**